client-runtime/src/main/java/com/microsoft/rest/v2/http/NettyClient.java

```diff
@@ -353,98 +360,206 @@ public void onComplete() {
+                return;
+            }
+            if (data instanceof LastHttpContent) {
+                done = true;
+            }
+            if (cancelled) {
+                data.release();
             } else {
-                chunksRequested = BackpressureHelper.addCap(chunksRequested, l);
+                queue.offer(data);
+                drain();
+            }
+        }
 
-                if (!draining) {
-                    try {
-                        draining = true;
-                        while (!queuedContent.isEmpty() && chunksRequested > 0 && !isCanceled) {
-                            emitContent(queuedContent.remove());
-                        }
+        void chunkCompleted() {
+            if (done) {
+                RxJavaPlugins.onError(new IllegalStateException("chunk completion event occurred after done is true"));
+                return;
+            }
+            chunkRequested.set(false);
+            drain();
+        }
 
-                        if (chunksRequested > 0 && !isCanceled) {
-                            handlerSubscription.request(chunksRequested);
-                        }
-                    } finally {
-                        draining = false;
-                    }
-                }
+        void onError(Throwable cause) {
+            if (done) {
+                RxJavaPlugins.onError(cause);
+            }
+            done = true;
+            err = cause;
+            drain();
+        }
+        
+        void channelInactive() {
+            if (!done) {
+                done = true;
+                err = new IOException("channel inactive");
+                drain();
             }
         }
 
-        @Override
-        public void cancel() {
-            isCanceled = true;
-            handlerSubscription.cancel();
+        //
+        // PRIVATE METHODS
+        //
 
-            currentEventLoop.execute(() -> {
-                while (!queuedContent.isEmpty()) {
-                    queuedContent.remove().release();
+        private void requestChunkOfByteBufsFromUpstream() {
+            channelSubscription.request(1);
+        }
+        
+        private void drain() {
+            if (wip.getAndIncrement() == 0) {
+                // need to check cancelled even if there are no requests
+                if (cancelled) {
+                    releaseQueue();
+                    return;
                 }
-            });
+                int missed = 1;
+                long r = requested.get();
+                while (true) {
+                    long e = 0;
+                    while (e != r) {
+                        // Note that an error can shortcut the emission of content that is currently on the queue.
+                        // This is probably desirable generally because it prevents work that being done downstream 
+                        // that might be thrown away anyway due to the error. 
+                        // TODO discuss with project members
```

Comment:
I think the behavior you've implemented is the preferred one-- if an error occurred in the producer, release the queued ByteBufs instead of propagating them, and propagate the error instead.

Reply:
good oh, I'll remove the todo line